From be9a0487360f497b37aa23f133005c3e00bdbc20 Mon Sep 17 00:00:00 2001
From: ReplayCoding <replaycoding@gmail.com>
Date: Sun, 24 Apr 2022 16:16:39 -0700
Subject: [PATCH] input: damage term when cancelling search using mouse

---
 input.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/input.c b/input.c
index 961b73ba..e55bf1ec 100644
--- a/input.c
+++ b/input.c
@@ -2227,6 +2227,7 @@ wl_pointer_button(void *data, struct wl_pointer *wl_pointer,
         break;
 
     case TERM_SURF_GRID: {
+        term_damage_view(term);
         search_cancel(term);
         urls_reset(term);
 
-- 
2.35.3

