From 3df1a8b969f8d380c4fbde8fcf2f4002ae976d2d Mon Sep 17 00:00:00 2001
From: ReplayCoding <replaycoding@gmail.com>
Date: Tue, 14 Dec 2021 10:59:52 -0800
Subject: [PATCH] fix compilation error in swaybar

---
 swaybar/tray/dbusmenu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/swaybar/tray/dbusmenu.c b/swaybar/tray/dbusmenu.c
index e4a6a9f6..968a55c6 100644
--- a/swaybar/tray/dbusmenu.c
+++ b/swaybar/tray/dbusmenu.c
@@ -546,7 +546,7 @@ static void swaybar_dbusmenu_draw_menu(struct swaybar_dbusmenu_menu *menu,
 		menu->dbusmenu->drawing = false;
 		return;
 	}
-	int surface_x, surface_y, surface_width, surface_height;
+	int surface_x, surface_y, surface_width, surface_height = {0};
 	draw_menu_items(cairo, menu, &surface_x, &surface_y, &surface_width,
 			&surface_height, open);
 
-- 
2.34.0

