From 3f7877aa846c03c6e687c397fe5a5f8db3484020 Mon Sep 17 00:00:00 2001
From: ReplayCoding <replaycoding@gmail.com>
Date: Tue, 21 Feb 2023 16:18:49 -0800
Subject: [PATCH 3/3] Disable updater

---
 src/components/VencordSettings/index.tsx | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/components/VencordSettings/index.tsx b/src/components/VencordSettings/index.tsx
index c15944c..ddd6250 100644
--- a/src/components/VencordSettings/index.tsx
+++ b/src/components/VencordSettings/index.tsx
@@ -50,7 +50,7 @@ const SettingsTabs: Record<string, SettingsTab> = {
     VencordSettingsSync: { name: "Backup & Restore", component: () => <BackupRestoreTab /> }
 };
 
-if (!IS_WEB) SettingsTabs.VencordUpdater.component = () => Updater && <Updater />;
+// if (!IS_WEB) SettingsTabs.VencordUpdater.component = () => Updater && <Updater />;
 
 function Settings(props: SettingsProps) {
     const { tab = "VencordSettings" } = props;
-- 
2.39.2

